File: questPresentation/src/main/java/ru/nekit/android/qls/setupWizard/QuestSetupWizard.java
package ru.nekit.android.qls.setupWizard;

import android.annotation.SuppressLint;
import android.app.admin.DevicePolicyManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

import ru.nekit.android.qls.EventBus;
import ru.nekit.android.qls.deviceAdminSupport.DeviceAdminComponent;
import ru.nekit.android.qls.lockScreen.LockScreen;
import ru.nekit.android.qls.pupil.PhoneContact;
import ru.nekit.android.qls.pupil.Pupil;
import ru.nekit.android.qls.pupil.PupilSex;
import ru.nekit.android.qls.quest.qtp.QuestTrainingProgramComplexity;
import ru.nekit.android.qls.setupWizard.view.QuestSetupWizardActivity;
import ru.nekit.android.qls.utils.PhoneManager;

import static ru.nekit.android.qls.setupWizard.QuestSetupWizard.QuestSetupWizardStep.SETTINGS;
import static ru.nekit.android.qls.setupWizard.QuestSetupWizard.QuestSetupWizardStep.START;
import static ru.nekit.android.qls.setupWizard.QuestSetupWizard.StepFlag.CAN_BE_RESET_AFTER_SET;
import static ru.nekit.android.qls.setupWizard.QuestSetupWizard.StepFlag.SETTINGS_PARENT;
import static ru.nekit.android.qls.setupWizard.QuestSetupWizard.StepFlag.SETUP_WIZARD_PARENT;

public class QuestSetupWizard extends BaseSetupWizard {

    private static final String NAME = "questSetupWizard";

    @SuppressLint("StaticFieldLeak")
    private static QuestSetupWizard instance;

    private final EventBus mEventBus;

    private QuestSetupWizard(@NonNull Context context) {
        super(context);
        mEventBus = new EventBus(context);
    }

    public static QuestSetupWizard getInstance(@NonNull Context context) {
        if (instance == null) {
            instance = new QuestSetupWizard(context);
        }
        return instance;
    }

    private boolean overlayPermissionIsSet() {
        return Build.VERSION.SDK_INT < Build.VERSION_CODES.M || Settings.canDrawOverlays(mContext);
    }

    private boolean callPhonePermissionIsGranted() {
        return PhoneManager.callPhonePermissionIsGranted(mContext);
    }

    private boolean readContactsPermissionIsGranted() {
        return PhoneManager.readContactsPermissionIsGranted(mContext);
    }

    public boolean allPermissionsIsGranted() {
        boolean permissionIsGranted = true;
        if (PhoneManager.phoneIsAvailable(mContext)) {
            permissionIsGranted = callPhonePermissionIsGranted()
                    && readContactsPermissionIsGranted();
        }
        return permissionIsGranted && overlayPermissionIsSet();
    }

    public void start() {
        mContext.startActivity(getStartIntent(true));
    }

    public Intent getStartIntent(boolean newTask) {
        Intent intent = new Intent(mContext, QuestSetupWizardActivity.class);
        if (newTask) {
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        } else {
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            intent.setFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
        }
        return intent;
    }

    public boolean setupIsComplete() {
        return mSettingsStorage.setupWizardIsCompleted(NAME);
    }

    private boolean unlockPasswordIsSet() {
        return LockScreen.getUnlockSecret() != null;
    }

    private boolean knoxIsSupport() {
        PackageManager packageManager = mContext.getPackageManager();
        return packageManager.hasSystemFeature("com.sec.android.mdm");
    }

    public ComponentName getDeviceAdminComponent() {
        return new ComponentName(mContext, DeviceAdminComponent.class);
    }

    public boolean deviceAdminIsActive() {
        DevicePolicyManager devicePolicyManager = (DevicePolicyManager) mContext.getSystemService(Context.DEVICE_POLICY_SERVICE);
        return devicePolicyManager.isAdminActive(getDeviceAdminComponent());
    }

    private boolean pupilIsRegistered() {
        return getPupil() != null;
    }

    @Nullable
    public Pupil getPupil() {
        return mPupilManager.getCurrentPupil();
    }

    public void setPupilName(@NonNull String name) {
        Pupil pupil = getPupil();
        if (pupil == null) {
            pupil = new Pupil();
            pupil.name = name;
            mPupilManager.addIfAbleAndNotify(pupil, true, mEventBus);
        } else {
            pupil.name = name;
            mPupilManager.update(pupil);
        }
    }

    public void setPupilSex(@NonNull PupilSex sex) {
        Pupil pupil = getPupil();
        if (pupil != null) {
            pupil.sex = sex;
            mPupilManager.update(pupil);
        }
    }

    public void setPupilAvatar(@NonNull String value) {
        Pupil pupil = getPupil();
        if (pupil != null) {
            pupil.avatar = value;
            mPupilManager.update(pupil);
        }
    }

    public void setQTPComplexity(@NonNull QuestTrainingProgramComplexity complexity) {
        Pupil pupil = getPupil();
        if (pupil != null) {
            pupil.complexity = complexity;
            mPupilManager.update(pupil);
        }
    }

    @NonNull
    @Override
    protected QuestSetupWizardStep calculateNextStep() {
        QuestSetupWizardStep[] steps = QuestSetupWizardStep.values();
        if (setupIsCompleted()) {
            for (QuestSetupWizardStep step : steps) {
                int flags = step.getFlags();
                if ((flags & CAN_BE_RESET_AFTER_SET) != 0 && (flags & SETUP_WIZARD_PARENT) != 0) {
                    if (!step.setupIsComplete(this)) {
                        return step;
                    }
                }
            }
        } else {
            for (QuestSetupWizardStep step : steps) {
                if (mCurrentStep == null) {
                    return step;
                } else {
                    if ((step.getFlags() & SETUP_WIZARD_PARENT) != 0) {
                        if (!step.setupIsComplete(this)) {
                            return step;
                        }
                    }
                }
            }
        }
        return SETTINGS;
    }

    public void setIntroductionIsPresented(boolean value) {
        mSettingsStorage.setIntroductionIsPresented(value);
    }

    @Override
    protected String getName() {
        return NAME;
    }

    public void commitCurrentSetupStep(ISetupWizardStep value) {
        if (value == SETTINGS) {
            completeSetupWizard();
        }
        if (value != START) {
            startSetupWizard();
        }
        this.mCurrentStep = value;
    }

    public void setUnlockSecret(String value) {
        LockScreen.setUnlockSecret(value);
    }

    public boolean checkUnlockSecret(String password) {
        return LockScreen.tryToLogin(password);
    }

    public void setAllowContacts(List<PhoneContact> value) {
        Pupil pupil = getPupil();
        if (pupil != null) {
            pupil.setAllowContacts(value);
        }
        mPupilManager.update(pupil);
    }

    public List<PhoneContact> getPhoneContacts() {
        Pupil pupil = getPupil();
        if (pupil != null) {
            return pupil.getAllowContacts();
        }
        return null;
    }

    public void switchOff() {
        LockScreen.switchOff(mContext);
    }

    public boolean phoneIsAvailable() {
        return PhoneManager.phoneIsAvailable(mContext);
    }

    public boolean lockScreenIsActive() {
        return LockScreen.isActive(mContext);
    }

    public void setDeviceAdminRemovable(boolean value) {
        if (knoxIsSupport()) {
            /*EnterpriseLicenseManager enterpriseLicenseManager = EnterpriseLicenseManager.getInstance(mContext);
            enterpriseLicenseManager.activateLicense();
            EnterpriseDeviceManager edm = new EnterpriseDeviceManager(mContext);
            edm.setAdminRemovable(value);*/
        }
    }

    public enum QuestSetupWizardStep implements ISetupWizardStep {

        START(SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return true;
            }
        },
        SETUP_UNLOCK_SECRET(SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.unlockPasswordIsSet();
            }
        },
        DEVICE_ADMIN(SETUP_WIZARD_PARENT | CAN_BE_RESET_AFTER_SET) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.deviceAdminIsActive();
            }
        },
        /*SAMSUNG_ENTERPRISE(SETUP_WIZARD_PARENT | CAN_BE_RESET_AFTER_SET) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return false;
            }
        },*/
        OVERLAY_PERMISSION(SETUP_WIZARD_PARENT | CAN_BE_RESET_AFTER_SET) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.overlayPermissionIsSet();
            }
        },
        BIND_PARENT_CONTROL(SETTINGS_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                throw new UnsupportedOperationException();
            }
        },
        CHANGE_UNLOCK_SECRET(SETTINGS_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                throw new UnsupportedOperationException();
            }
        },
        CALL_PHONE_AND_READ_CONTACTS_PERMISSION(SETUP_WIZARD_PARENT | CAN_BE_RESET_AFTER_SET) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.phoneIsAvailable() && setupWizard.callPhonePermissionIsGranted();
            }
        },
        PUPIL_NAME(SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.pupilIsRegistered();
            }
        },
        PUPIL_SEX(SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.getPupil().sex != null;
            }
        },
        QTP_COMPLEXITY(SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.getPupil().complexity != null;
            }
        },
        PUPIL_AVATAR(SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return setupWizard.getPupil().avatar != null;
            }
        },
        SETUP_ALLOW_CONTACTS(SETTINGS_PARENT | SETUP_WIZARD_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                List<PhoneContact> phoneContacts = setupWizard.getPhoneContacts();
                return setupWizard.phoneIsAvailable()
                        && phoneContacts != null
                        && phoneContacts.size() >= 0;
            }
        },
        SETTINGS(SETTINGS_PARENT) {
            @Override
            public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard) {
                throw new UnsupportedOperationException();
            }
        };

        private int mFlags;

        QuestSetupWizardStep(int flags) {
            mFlags = flags;
        }

        @Override
        public int getFlags() {
            return mFlags;
        }

        @Override
        public boolean needLogin() {
            return !(this == SETUP_UNLOCK_SECRET || this == START);
        }

        @Override
        public boolean needInternetConnection() {
            return this == BIND_PARENT_CONTROL;
        }

        abstract public boolean setupIsComplete(@NonNull QuestSetupWizard setupWizard);
    }

    static class StepFlag {

        static final int SETUP_WIZARD_PARENT = 1;
        static final int SETTINGS_PARENT = 2;
        static final int CAN_BE_RESET_AFTER_SET = 4;

    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/setupWizard/QuestSetupWizard.kt
package ru.nekit.android.qls.setupWizard

import android.app.Activity
import android.app.admin.DevicePolicyManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.Intent.*
import android.os.Build
import android.provider.Settings
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.disposables.Disposable
import ru.nekit.android.domain.interactor.use
import ru.nekit.android.domain.model.Optional
import ru.nekit.android.qls.data.representation.getSkuId
import ru.nekit.android.qls.data.representation.getSkuType
import ru.nekit.android.qls.dependences.DependenciesProvider
import ru.nekit.android.qls.domain.model.PhoneContact
import ru.nekit.android.qls.domain.model.SKU
import ru.nekit.android.qls.domain.model.SKUPurchase
import ru.nekit.android.qls.domain.model.SessionType
import ru.nekit.android.qls.domain.useCases.*
import ru.nekit.android.qls.lockScreen.LockScreen
import ru.nekit.android.qls.setupWizard.QuestSetupWizard.QuestSetupWizardStep.SETTINGS
import ru.nekit.android.qls.setupWizard.QuestSetupWizard.QuestSetupWizardStep.START
import ru.nekit.android.qls.setupWizard.QuestSetupWizard.StepFlag.HAS_PERMISSION
import ru.nekit.android.qls.setupWizard.QuestSetupWizard.StepFlag.SETTINGS_PARENT
import ru.nekit.android.qls.setupWizard.QuestSetupWizard.StepFlag.SETUP_WIZARD_PARENT
import ru.nekit.android.qls.setupWizard.deviceAdmin.DeviceAdminComponent
import ru.nekit.android.qls.setupWizard.view.QuestSetupWizardActivity
import ru.nekit.android.qls.shared.model.Complexity
import ru.nekit.android.qls.shared.model.Pupil
import ru.nekit.android.qls.shared.model.PupilSex
import ru.nekit.android.utils.PhoneUtils
import ru.nekit.android.utils.SingletonHolder
import ru.nekit.android.utils.toSingle

class QuestSetupWizard private constructor(private val dependenciesProvider: DependenciesProvider) :
        BaseSetupWizard(dependenciesProvider) {

    val timeProvider get() = dependenciesProvider.getTimeProvider()
    private val billing get() = dependenciesProvider.getBilling()
    val phoneContacts get() = PhoneContactsUseCases.getPhoneContacts()
    val deviceAdminComponent get() = ComponentName(dependenciesProvider, DeviceAdminComponent::class.java)
    val pupil get() = PupilUseCases.getCurrentPupil()
    private val questSetupWizardRepository
        get() = dependenciesProvider.repositoryHolder.getSettingsRepository()

    private fun getPurchasedSubscription(body: (SKUPurchase?) -> Unit) =
            SKUUseCases.purchasedSubscription().use { it -> body(it.data) }

    fun initiateSKUPurchaseFlow(activity: Activity, sku: SKU) = getPurchasedSubscription { oldSku ->
        billing.initiatePurchaseFlow(activity, sku.getSkuId(dependenciesProvider),
                oldSku?.sku?.getSkuId(dependenciesProvider), sku.skuType.getSkuType())
    }

    private fun overlayPermissionIsSet() =
            Build.VERSION.SDK_INT < Build.VERSION_CODES.M ||
                    Settings.canDrawOverlays(dependenciesProvider)

    private fun allPermissionsIsGranted(): Boolean = QuestSetupWizardStep.values().all {
        it.permissionIsSet(this)
    }

    fun addPhoneContact(value: PhoneContact): Completable =
            PhoneContactsUseCases.addPhoneContact(value)

    fun removePhoneContact(value: PhoneContact): Completable =
            PhoneContactsUseCases.removePhoneContact(value)

    private fun unlockPasswordIsSet(): Single<Boolean> = UnlockSecretUseCases.unlockSecretIsSet()

    /*
    private fun knoxIsSupport(): Boolean {
        val packageManager = context.packageManager
        return packageManager.hasSystemFeature("com.sec.android.mdm")
    }
    */

    fun deviceAdminPermissionsIsActive(): Boolean =
            (dependenciesProvider.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager).isAdminActive(deviceAdminComponent)

    /*fun setPupilName(name: String): Single<Boolean> =
            updatePupilParameter { it.name = name }*/

    fun setPupilSex(sex: PupilSex): Single<Boolean> =
            updatePupilParameter { it.sex = sex }

    fun setPupilAvatar(avatar: String): Single<Boolean> =
            updatePupilParameter { it.avatar = avatar }

    fun setQTPComplexity(complexity: Complexity?): Single<Boolean> =
            updatePupilParameter { it.complexity = complexity }

    private fun updatePupilParameter(body: (Pupil) -> Unit): Single<Boolean> =
            PupilUseCases.updatePupil(body)

    override fun calculateNextStep(): Single<ISetupWizardStep> =
            SetupWizardUseCases.setupIsComplete().build().flatMapPublisher { isComplete ->
                Flowable.fromIterable(QuestSetupWizardStep.values().toList()).concatMap { step ->
                    if (isComplete) {
                        val flags = step.flags
                        if (flags and HAS_PERMISSION != 0 && flags and SETUP_WIZARD_PARENT != 0)
                            step.stepIsComplete(this).map { Optional(if (it) null else step) }.toFlowable()
                        else
                            Flowable.just(Optional())
                    } else
                        if (currentStep == null)
                            Flowable.just(Optional(step))
                        else
                            if (step.flags and SETUP_WIZARD_PARENT != 0)
                                step.stepIsComplete(this).map { Optional(if (it) null else step) }.toFlowable()
                            else
                                Flowable.just(Optional())
                }
            }.filter { it.isNotEmpty() }.map { it.nonNullData }.first(SETTINGS).cast(ISetupWizardStep::class.java)

    override fun needLogin(step: ISetupWizardStep): Single<Boolean> =
            if (step.needLogin())
                SessionUseCases.checkSessionValidation(SessionType.SETUP_WIZARD).map { !it }
            else false.toSingle()

    fun commitCurrentSetupStep(value: ISetupWizardStep) {
        if (value == SETTINGS) {
            completeSetupWizard()
        }
        if (value != START) {
            startSetupWizard()
        }
        currentStep = value
    }

    override fun setupIsStart(body: (Boolean) -> Unit) =
            SetupWizardUseCases.setupIsStart(body)

    override fun setupIsComplete(body: (Boolean) -> Unit) =
            SetupWizardUseCases.setupIsComplete {
                body(it && allPermissionsIsGranted())
            }

    override fun completeSetupWizard() {
        SetupWizardUseCases.completeSetupWizard()
        LockScreen.getInstance().startForSetupWizard()
        QuestTrainingProgramUseCases.createQuestTrainingProgram()
    }

    override fun startSetupWizard() {
        SetupWizardUseCases.startSetupWizard()
        LockScreen.getInstance().startForSetupWizard()
    }

    fun setUnlockSecret(value: String): Completable = UnlockSecretUseCases.setUnlockSecret(value)

    fun checkUnlockSecret(password: String, body: (Boolean) -> Unit): Unit =
            UnlockSecretUseCases.checkUnlockSecret(password, body)

    fun phoneIsAvailable(): Boolean {
        return PhoneUtils.phoneIsAvailable(dependenciesProvider)
    }

    fun lockScreenIsSwitchedOn(body: (Boolean) -> Unit) = LockScreen.getInstance().isSwitchedOn(body)

    fun createBindCode(): Single<String> =
            pupil.flatMap {
                Single.just(if (it.isEmpty()) "" else
                    with(it.nonNullData) {
                        String.format(BIND_CODE_PATTERN, uuid, name)
                    }
                )
            }

    fun createPupil(pupilName: String, pupilSex: PupilSex): Single<Boolean> =
            PupilUseCases.createPupil(pupilName, pupilSex)

    fun pause() = LockScreen.getInstance().pause()

    /*
    fun setDeviceAdminRemovable(visibility: Boolean) {
        if (knoxIsSupport()) {
            /*EnterpriseLicenseManager enterpriseLicenseManager = EnterpriseLicenseManager.getInstance(context);
            enterpriseLicenseManager.activateLicense();
            EnterpriseDeviceManager edm = new EnterpriseDeviceManager(context);
            edm.setAdminRemovable(visibility);*/
        }
    }
    */

    fun deviceAdminIsSet(body: (Boolean) -> Unit): Disposable =
            stepIsComplete(QuestSetupWizard.QuestSetupWizardStep.DEVICE_ADMIN, body)

    fun stepIsComplete(step: QuestSetupWizardStep, body: (Boolean) -> Unit): Disposable =
            step.stepIsComplete(this).subscribe(body)

    private val useQTPComplexity: Boolean
        get() = !questSetupWizardRepository.useSingleQTP && questSetupWizardRepository.useQTPComplexity

    enum class QuestSetupWizardStep(val flags: Int) : ISetupWizardStep {

        START(SETUP_WIZARD_PARENT),
        SET_UNLOCK_SECRET(SETUP_WIZARD_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                return setupWizard.unlockPasswordIsSet()
            }
        },
        DEVICE_ADMIN(SETUP_WIZARD_PARENT or HAS_PERMISSION) {
            override fun permissionIsSet(setupWizard: QuestSetupWizard): Boolean {
                return setupWizard.deviceAdminPermissionsIsActive()
            }
        },
        /*SAMSUNG_ENTERPRISE(SETUP_WIZARD_PARENT | HAS_PERMISSION) {
            @Override
            public boolean stepIsComplete(@NonNull QuestSetupWizard setupWizard) {
                return false;
            }
        },*/
        OVERLAY_PERMISSION(SETUP_WIZARD_PARENT or HAS_PERMISSION) {
            override fun permissionIsSet(setupWizard: QuestSetupWizard): Boolean {
                return setupWizard.overlayPermissionIsSet()
            }
        },
        BIND_PARENT_CONTROL(SETTINGS_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                throw UnsupportedOperationException()
            }
        },
        CHANGE_UNLOCK_SECRET(SETTINGS_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                throw UnsupportedOperationException()
            }
        },
        CALL_PHONE_AND_READ_CONTACTS_PERMISSION(SETUP_WIZARD_PARENT or HAS_PERMISSION) {
            override fun permissionIsSet(setupWizard: QuestSetupWizard): Boolean {
                return if (setupWizard.phoneIsAvailable()) setupWizard.callPhonePermissionIsGranted()
                        && setupWizard.readContactsPermissionIsGranted()
                else true
            }
        },
        PUPIL_NAME_AND_SEX(SETUP_WIZARD_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                return setupWizard.pupil.map {
                    it.isNotEmpty() && it.nonNullData.name?.isNotEmpty() != null
                }
            }

        },
        /*PUPIL_SEX(SETUP_WIZARD_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                return setupWizard.pupil.map {
                    it.data?.sex != null
                }
            }
        },*/
        QTP_COMPLEXITY(SETUP_WIZARD_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> =
                    setupWizard.pupil.map {
                        if (setupWizard.useQTPComplexity)
                            it.data?.complexity != null
                        else
                            true
                    }
        },
        PUPIL_AVATAR(SETUP_WIZARD_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                return setupWizard.pupil.map { it.data?.avatar != null }
            }
        },
        SETUP_PHONE_CONTACTS(SETTINGS_PARENT or SETUP_WIZARD_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> =
                    setupWizard.phoneContacts.map {
                        it.isNotEmpty() && setupWizard.phoneIsAvailable()
                    }
        },
        VOICE_RECORD(SETTINGS_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                throw UnsupportedOperationException()
            }
        },
        SUBSCRIBES(SETTINGS_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                throw UnsupportedOperationException()
            }
        },
        SETTINGS(SETTINGS_PARENT) {
            override fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> {
                throw UnsupportedOperationException()
            }
        };

        override fun needLogin() = !(this == SET_UNLOCK_SECRET || this == START)

        override fun needInternetConnection(): Boolean {
            return this == BIND_PARENT_CONTROL
        }

        open fun stepIsComplete(setupWizard: QuestSetupWizard): Single<Boolean> =
                (if (flags and HAS_PERMISSION != 0)
                    permissionIsSet(setupWizard)
                else
                    true).toSingle()

        open fun permissionIsSet(setupWizard: QuestSetupWizard): Boolean = true

    }

    internal object StepFlag {

        const val SETUP_WIZARD_PARENT = 1
        const val SETTINGS_PARENT = 2
        //its can be reset or unset
        const val HAS_PERMISSION = 4

    }

    companion object : SingletonHolder<QuestSetupWizard, DependenciesProvider>(::QuestSetupWizard) {

        fun start(context: Context) = context.startActivity(getStartIntent(context, true))

        fun getStartIntent(context: Context, newTask: Boolean): Intent =
                Intent(context, QuestSetupWizardActivity::class.java).apply {
                    if (newTask) {
                        flags = FLAG_ACTIVITY_NEW_TASK
                    } else {
                        flags = FLAG_ACTIVITY_CLEAR_TOP
                        flags = FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS
                        flags = FLAG_ACTIVITY_CLEAR_TASK or FLAG_ACTIVITY_NEW_TASK
                    }
                }
    }

}